plugins/amqp_compressed_proof_plugin/amqp_compressed_proof_plugin.cpp

```diff
@@ -0,0 +1,393 @@
+         fc::raw::unpack(file, count);
+         while(count.value--) {
+            compressed_proof_generator_impl::reversible_action_entries entry;
+            fc::raw::unpack(file, entry);
+            my->reversible_action_entries_index.emplace(entry);
+         }
+      } FC_RETHROW_EXCEPTIONS(error, "Failed to load compressed merkle generator reversible state file ${f}", ("f", (fc::path)*my->reversible_path));
+   }
+   else {
+      boost::filesystem::ofstream s(*my->reversible_path);
+      EOS_ASSERT(s.good(), chain::plugin_config_exception, "Unable to create compressed merkle generator reversible state file ${f}", ("f", (fc::path)*my->reversible_path));
+   }
+   boost::system::error_code ec;
+   boost::filesystem::remove(*my->reversible_path, ec);
+}
+
+compressed_proof_generator::~compressed_proof_generator() {
+   if(my->reversible_path) {
+      try {
+         fc::unsigned_int num = my->reversible_action_entries_index.size();
+         fc::datastream<fc::cfile> file;
+         file.set_file_path(*my->reversible_path);
+         file.open("wb");
+         fc::raw::pack(file, num);
+         for(const auto& v : my->reversible_action_entries_index)
+            fc::raw::pack(file, v);
+      } FC_LOG_AND_DROP();
+   }
+}
+
+void compressed_proof_generator::on_applied_transaction(const chain::transaction_trace_ptr& trace) {
+   if(!trace->receipt)
+      return;
+   //only executed & delayed transaction traces would make it in to action_mroot
+   if(trace->receipt->status != chain::transaction_receipt::status_enum::executed &&
+      trace->receipt->status != chain::transaction_receipt::status_enum::delayed)
+      return;
+
+   if(my->is_onblock(trace))
+      my->onblock_trace.emplace(trace);
+   else
+      my->cached_traces[trace->id] = trace;
+}
+
+void compressed_proof_generator::on_accepted_block(const chain::block_state_ptr& bsp) {
+   std::set<action_entry> action_entries_this_block;
+
+   if(my->onblock_trace)
+      action_entries_this_block.emplace((*my->onblock_trace)->action_traces.front());
+   for(const chain::transaction_receipt& r : bsp->block->transactions) {
+      transaction_id_type id;
+      if(r.trx.contains<chain::transaction_id_type>())
+         id = r.trx.get<chain::transaction_id_type>();
+      else
+         id = r.trx.get<chain::packed_transaction>().id();
+      auto it = my->cached_traces.find(id);
+      EOS_ASSERT(it != my->cached_traces.end(), chain::misc_exception, "missing trace for transaction ${id}", ("id", id));
+      for(const chain::action_trace& at : it->second->action_traces)
+         action_entries_this_block.emplace(at);
+   }
+   my->cached_traces.clear();
+   my->onblock_trace.reset();
+
+   my->reversible_action_entries_index.emplace(generator_impl::reversible_action_entries{bsp->block_num, bsp->id, std::move(action_entries_this_block)});
+}
+
+void compressed_proof_generator::on_irreversible_block(const chain::block_state_ptr& bsp, const chain::controller& controller) {
```

Comment:
Do we always need to wait for irreversibility or can we just follow the database read mode and let the user decide?

Reply:
We don't here and neither does the witness plugin either. Is relying on the DB read mode really a good indication vs a separate config item though?

Reply:
The witness plugin is somewhat different.  Changing the point where it signs changes the meaning of the signature.  Signing before a block becomes irreversible is potentially dangerous.

Reply:
For the compressed proof plugin, decoupling the proof generation from on_accepted_block requires you to maintain significant extra state.  I don't thing it's worthwhile to make it configurable separately from the database read mode.